Synching my fork to Adafruit's updated repo

paynterf opened the pull request.

edgar-bonet commented:
I can't understand the purpose of this pull request. Was it submitted by accident?

paynterf commented:
Edgar,

The Arduino Wire library has known problems that cause it to hang forever
under certain edge conditions, because it's read() & write() functions are
blocking calls with no timeouts. The SBWIRE library was created to address
this issue by including timeouts on all blocking calls. This problem with
Wire has been known for at least a decade and hasn't been fixed, despite
multiple pull requests on the issue (see
https://www.fpaynter.com/2018/08/known-defect-in-arduino-i2c-code-causes-hangup-problems/).
Switching
to SBWIRE completely solved the problem.  Unfortunately, other I2C related
libraries like yours use the Wire library by default, and can't be used in
unmodified form in projects that have already switched to SBWIRE.

I incorporated an Adafruit DS3231 module into my project back in May 2018
(see
https://www.fpaynter.com/2018/05/integrating-time-memory-and-heading-capability/),
and wound up forking the RTC library and modifying RTClib.h/cpp to use the
SBWIRE library.  As a consequence of these mods I submitted a pull request
to the RTClib in February 2019 almost identical to the present one, but it
was never incorporated.  (my already modified local copy worked fine, so I
never followed up on the earlier pull request).  However, yesterday another
Adafruit RTC user with a similar hangup problem commented on my February
2019 'I2C Hangup' post, and this caused me to look at the RTClib repo
again.  When I looked, I discovered my original February 2019 pull request
had never been merged.

So, even though I personally had no reason to do so, I deleted my RTClib
folder, re-forked and re-cloned the current Adafruit RTClib, re-modified
and re-pushed my SBWIRE-related mods, and re-submitted the pull request for
the one-line change to RTClib.h and the 2-3 line change to RTClib.cpp, so
that other Adafruit RTC users with I2C hangup problems can switch to the
SBWIRE library by simply uncommenting one line in RTClib.h.

The bottom line though is this: The Wire library is fatally flawed and has
been for at least a decade.  Anyone who uses the Wire library for projects
that have to run for more than a few hours WILL experience I2C bus hangups.
The SBWIRE library fixes this issue, and should be the go-to library for
anything that uses the I2C bus.  In lieu of this, I2C-based libraries
should at least give their users an easy way to switch from Wire.h to
SBWIRE.h.  The REAL solution would be to get the @#$^$#& Arduino Wire
library maintainers to actually merge one of the many PRs submitted to them
over the last decade to fix the underlying hangup problems, but all PRs to
date have been rejected for one stupid reason or another.

I'm sure this was MUCH more than you ever wanted to hear about the history
of this pull request ;-)

I saw your other email/comment about why clang barfed, made the required
fixes, and committed them to my fork repo.  Then it was rejected again,
presumably because I didn't have a space between '//' and the rest of the
comment text.

Just as a totally unsolicited observation, you guys seem to be going the
way of the Arduino Wire library maintainers - making it easier on
yourselves, but harder on potential contributors by imposing things like
arbitrary 'clang' format checkers and not considering anything that doesn't
pass all the checks.  I personally have spent many hours now trying to
improve your library and help you sell more I2C-based RTC products, and you
have probably spent WAY more time with this issue than if you had simply
made the formatting mods yourself.  The Arduino Wire library maintainers
have so well protected their code from formatting problems that it has
completely ossified - and it seems to me you guys are well on the way to
making the same mistake.

Regards,

Frank




On Sun, Apr 26, 2020 at 5:00 AM Edgar Bonet <notifications@github.com>
wrote:

> I can't understand the purpose of this pull request. Was it submitted by
> accident?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/adafruit/RTClib/pull/159#issuecomment-619513282>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AA6T327RIECLHZOFXL4J7WDROPZ2JANCNFSM4MQ6JGGA>
> .
>


-- 
G.Frank Paynter, PhD
OSU ESL Research Scientist (ret)
EM Workbench LLC
614 638-6749 (cell)

siddacious commented:
This seems like it was created by mistake. If not, please explain what you were trying to do and I can re-open it if appropriate. 
**edit**: from your explanation, it appears you were trying to do what you've submitted in another PR. If that's not the case, please explain and we can re-open.

I'm assuming you were attempting to sync your fork with this repo. Here is a 'softer' approach:
https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/syncing-a-fork

If you have un-merged commits on master, you may wish to cherry pick them to a branch and take a more 'nuclear' approach:
https://gist.github.com/glennblock/1974465

paynterf commented:
forget it - you guys can make the changes or not.  If you don't, your customers will continue to suffer.  I have wasted way too much of my time over a period of two friggin years trying to get this one simple change merged into your library. It's not for my benefit, as I have already made the changes to my local copy.

Thanks for the links but I have no interest in becoming a Git bash expert.

siddacious commented:
First things first, @edgar-bonet needn't be saddled with answering for `clang-format` and how we (Adafruit) are maintaining this library. He is a valued contributor but he isn't responsible for the things you're taking issue with.

@paynterf regarding your feedback about automated formatting checks being a PITA to deal with, I agree that it's obnoxious if you're trying to correct them manually. I posted a comment in 
https://github.com/adafruit/RTClib/pull/160
suggesting to let `clang-format` do the formatting for you. Rest assured it will not change the behavior, only the formatting.

As for the necessity of automated formatting and compilation checks, it's a necessity of maintaining _hundreds_ of repositories. Enforcing a widely accepted formatting standard prevents unnecessary thrashing of non-functional parts of the code.

We're going to update the guide on library automation to include clang format and GitHub actions. I'll also make an effort to ensure that the information there is clear to contributors such as yourself, as to avoid the pain points you're running into.

It's Sunday morning and my 'day off' so I'll quickly comment on your issues with the Wire library: It's the default for Arduino and as such it should be the default for any libraries that want to seamlessly work with Arduino. Doing otherwise is a non-starter. It sounds like you dismissed the Arduino maintainer's requests for changes as a 'penmanship contest' which is your prerogative, but doing so without a fair consideration of why they might want those changes isn't going to do anything to get your code merged. For all I know they told you to stick your code where the sun don't shine, so I'm not going to speak for them, but as a maintainer of hundreds of projects, I appreciate it when contributors take the time to try and understand _why_ we're doing the things we do.

I value your and every other contributor's efforts to help our projects. The contributions may not always align with our goals for the projects, but the beauty of open source is that 'fork' button at the top right of the github page.
